---
title: "Unit 4: Scientific Questions and Statistical Models"
subtitle: "Correlation"
html-math-method:
  method: mathjax
  url: "https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"
format:
  clean-revealjs:
    self-contained: true
    chalkboard: false
    standalone: true
    slide-number: true
    show-slide-number: print
code-annotations: hover
bibliography: QMLS_Bibliography.bib
csl: evolution.csl
---

```{r}
#| label: setup
#| echo: false

library(tidyverse)
library(ggtext)
library(patchwork)
library(ggExtra)

theme_set(theme_classic(base_size = 20))
```


# Housekeeping {background-color="#40666e"}

- Three-week Check In: complete before Thursday (if you choose to)
- PS 3 due Thursday
    - PS keys available on Canvas
- Progress Check 1 next week
    - Tuesday: Questions and review (feel free to submit ahead)
    - Thursday: In class working time


# Questions {background-color="#f8c471"}


# What kinds of questions do we want to answer as scientists? {background-color="#40666e"}


## Scientific questions

- Is there an effect (e.g., mean differences)?
- How large is the effect?
- (How large) is the association?
- (How well) can I predict a quantity, group, or classification of interest?
- Many others


## Remember the Goal of Statistics

> "a way of taming **uncertainty**, of turning raw **data** into arguments that can resolve profound questions" [@Amabile1989-pf]

- The statistical analyses that you carry out are *models*.
- Inference depends on evaluating the relative support for different models.
    - Model where there is *no* effect ("Null")
    - Model where there is an effect ("Alternative(s)")
- Uncertainty is always present


## Example: Species richness in Missouri lakes

> Does the number of phytoplankton species change with agricultural intensity around lakes?

:::: {.columns}

::: {.column width="40%"}
![](https://i.imgur.com/GoOM9bb.png){fig-align="center"}
:::

::: {.column width="60%"}
- There is no relationship between species richness and agricultural intensity
- There is a positive relationship
- There is a negative relationship
:::

::::


## Example: Genetics of corn oil content

> Which set of loci best predicts oil content in corn kernels?

:::: {.columns}

::: {.column width="40%"}
![](https://i.imgur.com/w6z5m2e.jpg){fig-align="center"}
:::

::: {.column width="60%"}
- Heritability of corn oil content is 0
- Different possible combinations of SNPs
- All additive
- Gene-gene interactions
:::

::::


## Example: Canid relationships

> What are the phylogenetic relationships between dog breeds & other canids?

:::: {.columns}

::: {.column width="60%"}
![](https://i.imgur.com/spdaSkP.jpg){fig-align="center"}
:::

::: {.column width="40%"}
- Simultaneous & rapid radiation
- Single origin of dogs
- Multiple origins
- Ancient breeds are monophyletic
- Every possible tree....
:::

::::


## Example: Predict the species of a fossil

> Is this bone more like a chimpanzee, australopith, Neandertal, or modern human?

:::: {.columns}

::: {.column width="40%"}
![](https://i.imgur.com/cQ1Kl1v.png){fig-align="center"}
:::

::: {.column width="60%"}
- What is the percent correct classification of known fossils?
- What are the relative probabilities of new fossil being in one of four taxonomic groups?
:::

::::


## Example: Growth models in Alligators?

![](https://i.imgur.com/scZSjNE.png){fig-align="center"}

$$Determinate:~Length \left(Age\right) = \frac{\theta_1}{1 + \exp((\theta_2 - Age) / \theta_3)}$$

$$Indeterminate:~Length\left(Age\right) = \theta_0 + \theta_1 Age + \theta_2 Age^2$$


## Your Own Experiences


## Models and Inference

- You want to use data to answer questions.
- Make *inferences* using *models* with data.

The method you use for inference is part of the model.

- Different sets of assumptions
- Many will give very similar answers

:::{.center}
"clarifying the mapping from assumptions to inference to decision"^[https://statmodeling.stat.columbia.edu/2024/12/28/bayesian-inference-isnt-just-about-getting-the-answer-its-also-about-clarifying-the-mapping-from-assumptions-to-inference-to-decision/]
:::


## {background-image="https://i.imgur.com/kgPmw0b.jpg" background-size="100%"}


## Example: Wing Dimorphic Crickets

> Do long-winged crickets have a higher resting metabolic rate than short-winged crickets?

:::: {.columns}

::: {.column width="40%"}
![](https://i.imgur.com/kgPmw0b.jpg){fig-align="center"}
:::

::: {.column width="60%"}
Controlling for body mass,

- Metabolic rates do not differ
- Average metabolic rate of long-wings is higher
- Average  metabolic rate of long-wings is lower
:::

::::


## Observe some data

```{r}
CD <- read_csv("https://raw.githubusercontent.com/QM-Life-Sci/RDABR/refs/heads/main/data/Cricket_Metabolic_MassCor.csv",
               show_col_types = FALSE) |> 
  mutate(Wing_Morph = if_else(Wing_Morph == "LW", "Long Wing", "Short Wing"))
```

```{r}
#| echo: true

glimpse(CD)

CD |>
  group_by(Wing_Morph) |>
  summarise(mean = mean(CO2_resid))
```


## Observe data

```{r}
P <- CD |>
  ggplot(aes(Wing_Morph, CO2_resid)) +
  geom_point(position = position_jitter(width = 0.3), alpha = 0.25) +
  stat_summary(fun = mean, geom = "point", size = 4, color = "firebrick4") +
  labs(x = "Wing Morph", y = "Mass-Specific CO~2~") +
  theme(axis.title = element_markdown())
print(P)
```


## Do long-winged crickets have a higher resting metabolic rate than short-winged crickets?

- What is the probability the two groups have the same mean?
- What is the probability of observing a difference like this from sampling error only?
- What is the probability these are two samples from the same distribution?
- How well can you predict metabolic rate from wing morph?


## Could the observed difference be sampling error?

```{r}
print(P)
```


## How large a difference do you get from sampling error?

1. Assume long-winged and short-winged crickets share a distribution
2. Put all values from both morphs in one group
3. Randomly assign Long Wing and Short Wing labels to observed values of `CO2_resid`
4. Repeat over and over


## How large a difference do you get from sampling error?

```{r}
CD$Group <- "ALL"

CD |>
  ggplot(aes(Group, CO2_resid)) +
  geom_point(position = position_jitter(width = 0.3), alpha = 0.25) +
  stat_summary(fun = mean, geom = "point", size = 4, color = "firebrick4") +
  labs(x = "Wing Morph", y = "Mass-Specific CO~2~") +
  theme(axis.title = element_markdown())
  
```


## How large a difference do you get from sampling error?

```{r}
#| fig-align: center

CD$WM_Shuffle <- sample(CD$Wing_Morph)

CD |>
  group_by(WM_Shuffle) |>
  summarise(Mean = mean(CO2_resid))

CD |>
  ggplot(aes(WM_Shuffle, CO2_resid)) +
  geom_point(position = position_jitter(width = 0.3), alpha = 0.25) +
  stat_summary(fun = mean, geom = "point", size = 4, color = "firebrick4")  +
  labs(x = "Wing Morph", y = "Mass-Specific CO~2~") +
  theme(axis.title = element_markdown())
```


## How big of a difference do you get from sampling error?

```{r}
#| fig-align: center

CD$WM_Shuffle <- sample(CD$Wing_Morph)

CD |>
  group_by(WM_Shuffle) |>
  summarise(Mean = mean(CO2_resid))

CD |>
  ggplot(aes(WM_Shuffle, CO2_resid)) +
  geom_point(position = position_jitter(width = 0.3), alpha = 0.25) +
  stat_summary(fun = mean, geom = "point", size = 4, color = "firebrick4") +
  labs(x = "Wing Morph", y = "Mass-Specific CO~2~") +
  theme(axis.title = element_markdown())  
```


## What do we expect to see when there is a meaningful difference between wing morphs?

```{r}
print(P)
```


## What do we expect to see when there is a meaningful difference between wing morphs? 

1. Assume corrected metabolic rate is normally distributed
2. Draw random values for long wings from a normal distribution with a *higher* mean
3. Draw random values for short wings from a normal distribution with a *lower* mean
4. Repeat over and over


## Simulate a difference using distributions

```{r}
#| echo: true

n.sw <- sum(CD$Wing_Morph == "Short Wing")
n.lw <- sum(CD$Wing_Morph == "Long Wing")

mean.sw <- -1
mean.lw <- 2

SS <- tibble(Wing_Morph = c(rep("Short Wing", n.sw), rep("Long Wing", n.lw)),
             Sim_CO2 = c(rnorm(n.sw, mean.sw), rnorm(n.lw, mean.lw)))

SS |> group_by(Wing_Morph) |>
  summarise(Mean = mean(Sim_CO2))
```


## Simulate a difference using distributions

```{r}
#| echo: false

SS |> group_by(Wing_Morph) |>
  summarise(Mean = mean(Sim_CO2))

SS |>
  ggplot(aes(Wing_Morph, Sim_CO2)) +
  geom_point(position = position_jitter(width = 0.3), alpha = 0.25) +
  stat_summary(fun = mean, geom = "point", size = 4, color = "firebrick4")  +
  labs(x = "Wing Morph", y = "Mass-Specific CO~2~") +
  theme(axis.title = element_markdown())  
  
```


## Simulate a difference using distributions

```{r}
mean.sw <- -1
mean.lw <- 2

SS <- tibble(Wing_Morph = c(rep("Short Wing", n.sw), rep("Long Wing", n.lw)),
             Sim_CO2 = c(rnorm(n.sw, mean.sw), rnorm(n.lw, mean.lw)))

SS |> group_by(Wing_Morph) |>
  summarise(Mean = mean(Sim_CO2))

SS |>
  ggplot(aes(Wing_Morph, Sim_CO2)) +
  geom_point(position = position_jitter(width = 0.3), alpha = 0.25) +
  stat_summary(fun = mean, geom = "point", size = 4, color = "firebrick4")  +
  labs(x = "Wing Morph", y = "Mass-Specific CO~2~") +
  theme(axis.title = element_markdown())
```


# Models {background-color="#40666e"}


## Models and Inference

> All models are wrong, but some are useful.

:::{.right}
George Box (1976; attributed)
:::

- If we knew the process exactly we would not need a model (e.g., points exactly on a line).

$$y = mx + b$$

- Often not a single unequivocably correct model.
- Some very wrong and some less wrong.
- Your role is to choose and defend a model, given your questions and the assumptions of the model.


## Types of data

*Continuous*

- Real number line (rounding okay)
- Bone lengths, age, oil content of corn
- Counts of species

*Categorical*

- Binomial: 0/1, land/water, etc.
- Treatment or observation groups (Ag/Non-Ag land use)
- Genotypes (Sticklebacks: *MM*, *Mm*, *mm*)


# Strength of Association {background-color="#40666e"}

Pairs of continuous variables


## Coloration of lion noses

:::: {.columns}

::: {.column width="30%"}
![](images/lion_nose.png)
:::

::: {.column width="70%"}
```{r}
#| fig-height: 8

LA <- read_csv("https://raw.githubusercontent.com/Middleton-Lab/abdData/refs/heads/main/inst/extdata/datasets/17/17e1LionAges.csv", show_col_types = FALSE) |> 
  rename(Age = age,
         `Proportion Black` = proportion.black)

ggplot(LA, aes(Age, `Proportion Black`)) +
  geom_point(size = 6, color = "navy") +
  scale_y_continuous(limits = c(0, 1)) +
  scale_x_continuous(breaks = 1:14) +
  theme_classic(base_size = 30)
```
:::

::::


## Coloration of lion noses

```{r}
p1 <- ggplot(LA, aes(x = Age, y = `Proportion Black`)) +
  geom_point(size = 4, color = "navy") +
  scale_y_continuous(limits = c(0, 1)) +
  scale_x_continuous(breaks = 1:14) +
  theme_classic(base_size = 20)

p2 <- ggplot(LA, aes(y = Age, x = `Proportion Black`)) +
  geom_point(size = 4, color = "navy") +
  scale_x_continuous(limits = c(0, 1)) +
  scale_y_continuous(breaks = 1:14) +
  theme_classic(base_size = 20)
p1 + p2
```

:::{.center}
Which plot is better?
:::


## (Pearson's) Correlation ($r$)

Strength and direction of association between two variables

- Equally true:
    - How does $x$ change as $y$ changes?
    - How does $y$ change as $x$ changes?
- Range: $-1 \leq r \leq 1$
- Null hypothesis: $r = 0$


## (Pearson's) Correlation ($r$)

Strength of association between two variables:

\begin{align*}
r &= \frac{{}\sum_{i=1}^{n} (x_i - \bar{x})(y_i - \bar{y})}
{\sqrt{\sum_{i=1}^{n} (x_i - \bar{x})^2  \sum_{i=1}^{n}(y_i - \bar{y})^2}}\\
    &= \frac{1}{n - 1} \sum_{i=1}^n \frac{x_i - \bar{x}}{s_x} \frac{y_i - \bar{y}}{s_y}\\
\end{align*}

where

- $(x_i - \bar{x})$ and $(y_i - \bar{y})$ are deviations from the mean
- $s_x$ and $s_y$ are standard deviations


## Assumptions

- $x$ and $y$ are paired
- Observations are independent
- Bivariate normality
    - Separately, $x$ and $y$ are normally distributed


## Bivariate normality

```{r}
#| echo: true
#| output-location: slide

library(ggExtra)
set.seed(43237497)

BVN <- tibble(x = rnorm(1e4, mean = 0, sd = 1),
              y = rnorm(1e4, mean = 0, sd = 1))

P <- ggplot(BVN, aes(x, y)) +
  geom_point(alpha = 0.25, color = "firebrick") +
  coord_equal()

ggMarginal(P, type = "histogram", fill = "firebrick")
```


## Correlation in R: `cor()`

Random bivariate normal data:

```{r}
#| echo: true

cor(BVN$x, BVN$y)
```

Lion nose coloration:

```{r}
#| echo: true

cor(LA$Age, LA$`Proportion Black`)

cor(LA$`Proportion Black`, LA$Age)
```


## References

::: {#refs}
:::

